src/install-script/installer.ts

```diff
@@ -0,0 +1,200 @@
+import * as archiver from "archiver";
+import { ChildProcess, exec, spawn } from "child_process";
+import { createWriteStream, ensureDir, remove } from "fs-extra";
+import { join as joinPath } from "path";
+
+export enum ScriptMessageType {
+  MINOR,
+  MAJOR,
+  ERROR,
+  COMPLETE
+}
+
+export enum InfoSource {
+  NPM = "npm",
+  VERDACCIO = "verdaccio",
+  OFFLINE_NPM = "offline-npm",
+  ARCHIVER = "archiver"
+}
+
+export interface IScriptMessage {
+  type: ScriptMessageType;
+  message: string;
+  source: InfoSource;
+}
+
+// Utility to make using callbacks with IScriptMessages easier
+const makeOutputter = (
+  callback: (data: IScriptMessage) => void,
+  type: ScriptMessageType
+) => (message: string, source: InfoSource = InfoSource.OFFLINE_NPM) => {
+  callback({
+    message,
+    source,
+    type
+  });
+};
+
+/**
+ * Installs an npm cache through Verdaccio and bundles the results in a zip file.
+ * @param packages the list of package names to install
+ * @param onInfo callback function for info messages
+ * @param onError callback function for error messages
+ * @param onComplete callback function for when the installation is complete
+ * @param outputPath the directory in which storage.zip should be placed
+ * @param verdaccioPath the directory in which the verdaccio source is located (usually node_modules)
+ */
+export const installCache = (
+  packages: string[],
+  onInfo: (data: IScriptMessage) => void,
+  onError: (data: IScriptMessage) => void,
+  onComplete: (data: IScriptMessage) => void,
+  outputPath: string = joinPath(__dirname, "out"),
+  verdaccioPath: string = joinPath(
+    __dirname,
+    "..",
+    "..",
+    "node_modules",
+    "verdaccio"
+  )
+): void => {
+  const outMinor = makeOutputter(onInfo, ScriptMessageType.MINOR);
+  const outMajor = makeOutputter(onInfo, ScriptMessageType.MAJOR);
+  const outError = makeOutputter(onError, ScriptMessageType.ERROR);
+  const outComplete = makeOutputter(onComplete, ScriptMessageType.COMPLETE);
```

Comment:
I think once you combine the three on* loggers into one, you can probably do away with these as well, and just pass the severity with the logs.

Reply:
I'm also dubious of the benefits of using this:

```typescript
outMinor(information);
```

...over this:

```typescript
onInfo(information, ScriptMessageType.Minor);
```

...especially if, by including the message type in `information`, it just becomes `onInfo(information);` again.